extract wal start location from tar when using pipe

pellcorp opened the pull request.

pellcorp commented:
Please don't merge yet, there is a bug.  I am adding tests and fixing the bug and then it can be reviewed, apologies.

pellcorp commented:
ok I have fixed the bug and added tests to make sure that what is in the metadata is the same as whats in the backup label.

pellcorp commented:
ok sorry missed a couple of pylint issues in the new test code, fixed now so should be ready to review.  I have run make test locally (in my vagrant env) and its all passed.

pellcorp commented:
sorry this was just rebasing on master, I had forgot to update my origin master figured you would prefer these MR to be up to date.  Anyway assuming all passes its ready to merge.

pellcorp commented:
I am looking at the failing test, it seems unrelated, but I will investigate.

pellcorp commented:
I can't reproduce this, I have run the tests multiple with python 3.7 and 3.5 and 3.6.  I am triggering another build in case this is a temporary issue with postgres

pellcorp commented:
Its all good now, please review and let me know if there are reasons why this won't be accepted, this change is pivotal to our backups against replica using archiving.  Otherwise wals that are still required get wiped.   I would prefer to use an official version of pghoard rather than a local fork too.  

If you have concerns about this causing issues, can I assuage them by making this behaviour configurable with additional enhancements?

pellcorp commented:
Any chance this will be merged or are you ignoring it hoping I will go away 😁

Ormod commented:
I'm traveling this week but I'll try to read this next week when I'm back at the office.

pellcorp commented:
By way of additional context, the issue I fixed in this MR occurred again in our production environment using an old version of pghoard.  I have changed some of the host names to avoid exposing details of my production environment publicly.

so here is an example of the start wal sequence issue happening on production.  So the metadata file has a different start-wal-sequence to the actual file.  
I created a extract-backup-label.py to be useful for diagnostics on unencrypted base backups.

```
postgres@p-025:~$ cat /data/postgresql/backup/p-025.pellcorp.com/cluster001/basebackup/2019-03-20_2.metadata
{"pg-version": "90514", "start-time": "2019-03-20T18:44:06.654737+00:00", "original-file-size": "66542393344", "compression-algorithm": "snappy", "start-wal-segment": "00000001000034DC000000AD"}
```

```
postgres@p-025:~$ cat /data/postgresql/backup/p-025.pellcorp.com/cluster001/basebackup/2019-03-20_2 | extract-backup-label.py
START WAL LOCATION: 34DC/A6412520 (file 00000001000034DC000000A6)
CHECKPOINT LOCATION: 34DC/ADFC0880
BACKUP METHOD: streamed
BACKUP FROM: standby
START TIME: 2019-03-20 19:37:07 CET
LABEL: pghoard_base_backup
```
And the wal file required to start the restore of this base backup has indeed been deleted by the old version of pghoard:

```
postgres@p-025:~$ ls /data/postgresql/backup/p-025.pellcorp.com/cluster001/xlog/00000001000034DC000000A6
ls: cannot access '/data/postgresql/backup/p-htz-025.adn.bitshift.technology/cluster001/xlog/00000001000034DC000000A6': No such file or directory
```

Whereas the one in the metadata is sticking around, but quite useless:
```
postgres@p-025:~$ ls /data/postgresql/backup/p-025.pellcorp.com/cluster001/xlog/00000001000034DC000000AD
/data/postgresql/backup/p-htz-025.adn.bitshift.technology/cluster001/xlog/00000001000034DC000000AD
```

pellcorp commented:
I should add that even in the test cases there were differences between the metadata and the base backups so the tests I added do prove my fix works

ojarva commented:
If you have a good idea on how to improve the test coverage, it is always a great idea to add those (up to some threshold, obviously), as regressions happen, and tests are a great way to prevent those.

pellcorp commented:
I did add tests for this change, well I adjusted the existing tests

pellcorp commented:
Any chance this will be merged?
